pkg/upload/js.go

```diff
@@ -108,6 +109,101 @@ func resolveSourceMapPaths(sourceMapPath string, outputPath string, logger log.L
 	return sourceMapPaths, nil
 }
 
+// Reads a JSON source map into memory. Required in order to check if the sourcesContent exists.
+func ReadSourceMap(path string, logger log.Logger) (map[string]interface{}, error) {
+	logger.Debug(fmt.Sprintf("Reading sourcemap %s", path))
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return map[string]interface{}{}, err
+	}
+	var sourceMapContents map[string]interface{}
+	err = json.Unmarshal(data, &sourceMapContents)
+	if err != nil {
+		return map[string]interface{}{}, err
+	}
+	return sourceMapContents, nil
+}
+
+func addSourcesContent(section map[string]interface{}, relativePath string, logger log.Logger) {
+	untypedSources, hasSources := section["sources"]
+	if !hasSources {
+		logger.Warn("Source map doesn't have required sources field")
+		return
+	}
+	sources, sourcesValid := untypedSources.([]interface{})
+	if !sourcesValid {
+		logger.Warn("Sources aren't an array")
+		return
+	}
+
+	// Skip if the sources content is already valid
+	untypedSourcesContent, hasSourcesContent := section["sourcesContent"]
+	if hasSourcesContent {
+		sourcesContent, sourcesContentValid := untypedSourcesContent.([]interface{})
+		if sourcesContentValid && len(sourcesContent) == len(sources) {
+			logger.Debug("SourcesContent is already populated")
+			return
+		}
+	}
+
+	var sourcesContent []*string
+	for _, sourcePath := range sources {
+		sourcePath, isString := sourcePath.(string)
+		// Skip null values
+		if !isString {
+			sourcesContent = append(sourcesContent, nil)
+			continue
+		}
+		// Handle the default webpack prefix in accordance with https://webpack.js.org/configuration/output/#outputdevtoolmodulefilenametemplate
+		sourcePath, isWebpack := strings.CutPrefix(sourcePath, "webpack://")
+		if isWebpack {
+			// Remove the namespace
+			firstSlash := strings.Index(sourcePath, "/")
+			if firstSlash != -1 && firstSlash+1 < len(sourcePath) {
+				sourcePath = sourcePath[firstSlash+1:]
+			}
+
+			// Skip virtual webpack files
+			if strings.Contains(sourcePath, "/webpack/") {
+				sourcesContent = append(sourcesContent, nil)
+				continue
+			}
+
+			// Remove any loaders
+			questionMark := strings.LastIndex(sourcePath, "?")
+			if questionMark-1 > 0 {
+				sourcePath = sourcePath[:questionMark-1]
+			}
+		}
+		if !path.IsAbs(sourcePath) {
+			sourcePath = path.Join(relativePath, sourcePath)
+		}
+		content, err := os.ReadFile(sourcePath)
+		if err != nil {
+			logger.Warn(fmt.Sprintf("Cannot read source file '%s' when searching relative to '%s'.", sourcePath, relativePath))
```

Comment:
`sourcePath` will have been amended to be absolute by here if it's relative, so I don't think the error message will read logically. How about just:


```suggestion
			logger.Warn(fmt.Sprintf("Cannot read referenced source file '%s'", sourcePath))
```